File: Java/src/com/practice/Hard/TrapRainWater.java
package com.practice.Hard;

import java.util.Stack;

public class TrapRainWater {
    public int trap(int[] height) {
        if (height.length == 0) return 0;
        int left = 0, right = height.length - 1;
        int area = 0, leftHeight = height[0], rightHeight = height[height.length - 1];
        while (left < right){
            if (height[left] < height[right]){
                left++;
                leftHeight = Math.max(leftHeight, height[left]);
                area += leftHeight - height[left];
            } else{
                right--;
                rightHeight = Math.max(rightHeight, height[right]);
                area += rightHeight - height[right];
            }
        }
        return area;
    }

    /**
     * passed oj
     * @param height
     * @return
     */
    public int trap_stack(int[] height) {
        int ans = 0;
        Stack<Integer>stk = new Stack<>();
        int i = 0;
        while(i < height.length){
            while(!stk.isEmpty() && height[i] > height[stk.peek()]){
                int oldStackHi = height[stk.peek()];
                stk.pop();

                if(stk.isEmpty()) break;

                int dis = (i- stk.peek() -1);
                int newHigherHi = Math.min(height[stk.peek()], height[i]);

                ans += dis * (newHigherHi - oldStackHi);
            }
            stk.push(i++);
        }
        return ans;
    }

}

File: Python/python3/leetcode/Hard/TrapRainWater.py
from typing import List

class TrapRainWater:

    def trap(self, height: List[int]) -> int:
        ans = 0
        i = 0
        stk = []
        while i < len(height):
            while len(stk) > 0 and height[i] > stk[-1]:
                oldStackHi = stk[-1]
                stk = stk[0:-2]

                if len(stk) == 0: break

                dist = i - stk[-1] - 1
                newHigherHi = min(stk[-1], height[i])
                ans += dist * (newHigherHi - oldStackHi)

            stk.append(i)
            i += 1

        return ans